keps/sig-cluster-lifecycle/20190227-artifact-generation.md

```diff
@@ -0,0 +1,197 @@
+
+## Motivation
+
+Today, the tooling used to build kubernetes packages is separate from kubernetes itself. This can cause downstream divergence, and cannot be easily maintained by the kubernetes community as a whole. 
+
+The primary goal of this proposal is to simplify the lifecycle management of packaging by co-locating the sources and tooling for generating packages into the kubernetes/kubernetes repository. 
+
+### Goals
+
+To create a maintainable and configurable method of building kubernetes artifacts supported and maintained by the community.
+
+- Move packaging tools out of kubernetes/release and into kubernetes/kubernetes repository
+- Enable customization for downstream packaging 
+    - Including the ability to bring their own source code/binaries and changelogs 
+    - Stop hard-coding the binary download URL to dl.k8s.io
+- Enable community governance over packaging of the project.
+- Tightly couple the versioning of packing and the kubernetes/kubernetes release.
+- Ensure that the same packages that are run in CI are the ones that are released.
+- Dedupe the generation of packages (THERE CAN BE ONLY ONE!) 
+- Have one config file that is the source of truth for packaging details for both .rpm and .deb packages.
+- Artifact generation is independent of source code compilation. 
+
+### Non-Goals
+
+- A change in the release process. see this [kep](https://github.com/kubernetes/enhancements/pull/843)
+    - For example, package signing and distribution are not covered in this document and are covered in the above KEP.
+
+- To fundamentally change the packages and their dependencies. All packages should remain consistent with what we build today. 
+
+## Proposal
+
+As outlined in the following issue:
+Base ticket: https://github.com/kubernetes/kubernetes/issues/71677
+Move all packaging into kubernetes/kubernetes repo and create the following 3 building blocks to support it.
+Create a configuration yaml file to enable the configuring of the built packages.
+Create a Go lang app to parse yaml config and submit simple commands to the container.
+Build inside a docker container based off FPM.
+
+### User Stories
+
+#### Story 1
+
+As a developer I would like to be able to install the entire set of alpha or beta (.deb/.rpm) packages to stand up a development cluster.  These packages should be equivalent to the future release artifacts.  
+
+#### Story 2
+
+As a third party vendor, I will be able to package kubernetes using the same tooling that kubernetes’ uses. 
+
+#### Story 3
+
+As a release manager I would like to enable simple and easy package generation that the community can manage. 
+
+### Implementation Details/Notes/Constraints [optional]
+
+- Use an fpm based container to build packages
+    - Nfpm was originally investigated but lacked features
+- Use yaml to configure
+- Create a go binary that can consume the yaml, build the right file structure and issue commands to the image to build the image.
```

Comment:
this might require a separate KEP

Reply:
I don't think so really, the go binary is the glue between the yaml config file and the docker container that builds it. without it you have two things that don't know about each other and we would be stuck shimming this into bazel maybe?

Reply:
I agree - we should not do a separate KEP for tooling we need to build packages.

Reply:
If the tool is single purpose, I agree.

Reply:
Let me know what you think. I tried to make this more clear @frapposelli